### PR TITLE
Add ReadyScheduled to DaemonSet status

### LINKED ISSUE #25605: Add Ready replica count to DaemonsetStatus
Readiness is as relevant to daemonsets as deployments, but while DeploymentStatus has "AvailableReplicas" i.e. the number of pods ready for at least minReadySeconds, DaemonsetStatus does not have an equivalent count of replicas that are healthy/ready. Not only can services point at daemonset pods, making readiness relevant, but readiness can also be used to indicate general health. I suggest adding "readyReplicas" to DaemonsetStatus.

The main use case should be for kubectl, but I also would like to add a metric to [kubernetes/kube-state-metrics](https://github.com/kubernetes/kube-state-metrics/) reflecting ready replicas of a daemonset, as it exists for deployments.

### PR DESCRIPTION
**Pull request by lukaszo**
Fixes #25605

cc @bgrant0607 @mikedanese 

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32781)

### CODE REVIEW

**Comment by smarterclayton** [comment]
@kubernetes/api-review-team for addition of `ReadyScheduled`.

The code looks correct, can you add two things - unit test to daemoncontroller_test.go, and add this to the existing daemonset e2e tests.

**Comment by lukaszo** [comment]
@smarterclayton I added unit test and added check to e2e tests. Please take a look

**Comment by lukaszo** [comment]
@kubernetes/daemonset ping. Please review

**Comment by mikedanese** [comment]
LGTM

**Comment by smarterclayton** [comment]
Changes LGTM to me as well, but still needs an API sign off.

**Comment by lukaszo** [comment]
@smarterclayton: @mikedanese is a member of @kubernetes/api-review-team and he gave lgtm

File: pkg/apis/extensions/types.go [hunk 1]
```diff
@@ -385,6 +385,9 @@ type DaemonSetStatus struct {
 	// DesiredNumberScheduled is the total number of nodes that should be running the daemon
 	// pod (including nodes correctly running the daemon pod).
 	DesiredNumberScheduled int32 `json:"desiredNumberScheduled"`
+
+	// The number of ready pods for this daemon set.
+	ReadyScheduled int32 `json:"readyScheduled"`
```
Comment (mikedanese) [inline pkg/apis/extensions/types.go]: NumberReady is probably a better name for this field.

**Review by mikedanese — CHANGES_REQUESTED** [review]
I did not give this an API review. I only reviewed the controller code. I would like @davidopp or @bgrant0607 to do a once over once you address my comments.

**Comment by bgrant0607** [comment]
cc @pwittrock @janetkuo @erictune @kargakis

**Comment by bgrant0607** [comment]
@lukaszo @mikedanese This is ready nodes or ready pods?

File: pkg/apis/extensions/types.go (hunk 1 above)
Comment (bgrant0607) [inline pkg/apis/extensions/types.go]: No field names contain the string "Number". I'd like to keep it that way. :-)

Comment (bgrant0607) [inline pkg/apis/extensions/types.go]: Ok, that's no longer true. But let's not introduce more occurrences.

Comment (mikedanese) [inline pkg/apis/extensions/types.go]: All the other fields in the DaemonSetStatus contain the string "Number".

**Comment by mikedanese** [comment]
@bgrant0607 This is the number of nodes that should be running the daemon pod and have one or more of the daemon pod running and ready.

**Comment by kargakis** [comment]
LGTM

File: pkg/apis/extensions/types.go [hunk 2]
```diff
@@ -385,6 +385,9 @@ type DaemonSetStatus struct {
 	// DesiredNumberScheduled is the total number of nodes that should be running the daemon
 	// pod (including nodes correctly running the daemon pod).
 	DesiredNumberScheduled int32 `json:"desiredNumberScheduled"`
+
+	// The number of ready pods for this daemon set.
```
Comment (lukaszo) [inline pkg/apis/extensions/types.go]: I updated the name and description

File: pkg/apis/extensions/types.go (hunk 1 above)
Comment (lukaszo) [inline pkg/apis/extensions/types.go]: I updated commits and changed to NumberReady

**Comment by lukaszo** [comment]
Ping. I need it for DaemonSet updates #31693

File: pkg/apis/extensions/types.go (hunk 1 above)
Comment (bgrant0607) [inline pkg/apis/extensions/types.go]: My own rule is that consistency trumps most other concerns, so any of NumberReady, ReadyNumberScheduled, or NumberReadyScheduled would be fine.

**Comment by lukaszo** [comment]
@mikedanese I changed name to NumberReady and @bgrant0607  accepts it. Do you need any more actions from me?

**Comment by lukaszo** [comment]
@k8s-bot gci gce e2e test this

**Comment by lukaszo** [comment]
Can someone restart the tests? Tests are passing for me locally.

**Comment by kargakis** [comment]
@lukaszo you need to rebase

**Comment by lukaszo** [comment]
@k8s-bot unit test this #33784.

**Comment by lukaszo** [comment]
@k8s-bot gce e2e test this

**Comment by lukaszo** [comment]
@k8s-bot gke e2e test this

**Comment by lukaszo** [comment]
@k8s-bot gci gke e2e test this

**Comment by lukaszo** [comment]
Ok, now all tests are passing. @kubernetes/daemonset PTAL

**Comment by lukaszo** [comment]
@mikedanese @kargakis can you add LGTM or do I need to change something more?

**Comment by kargakis** [comment]
lgtm

General question: how did we manage to mess the daemon status field names like that?:)

**Comment by lukaszo** [comment]
@k8s-bot gce e2e test this

**Comment by lukaszo** [comment]
@k8s-bot verify test this

**Comment by lukaszo** [comment]
@k8s-bot gke e2e test this

**Comment by lukaszo** [comment]
@kargakis  can you re-lgtm? I had to rebase again

File: pkg/controller/daemon/daemoncontroller.go [hunk 3]
```diff
@@ -580,6 +582,12 @@ func (dsc *DaemonSetsController) updateDaemonSetStatus(ds *extensions.DaemonSet)
 			desiredNumberScheduled++
 			if scheduled {
 				currentNumberScheduled++
+				// Sort the daemon pods by creation time, so the the oldest is first.
+				daemonPods, _ := nodeToDaemonPods[node.Name]
+				sort.Sort(podByCreationTimestamp(daemonPods))
```
Comment (kargakis) [inline pkg/controller/daemon/daemoncontroller.go]: Do we need to sort pods here?

Comment (lukaszo) [inline pkg/controller/daemon/daemoncontroller.go]: I'm using the same approach as here https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/daemon/daemoncontroller.go#L458-L460
If `dsc.getNodesToDaemonPods(ds)` returns more than one pod on node I want to get status of the youngest pod.

**Comment by spxtr** [comment]
@k8s-bot test this

**Comment by lukaszo** [comment]
@k8s-bot unit test this #34547

**Comment by lukaszo** [comment]
@k8s-bot gce e2e test this 
I think that it's this flake issue #33867

File: pkg/controller/daemon/daemoncontroller.go (hunk 3 above)
Comment (lukaszo) [inline pkg/controller/daemon/daemoncontroller.go]: Ah yes, the oldest. Sorry.

**Comment by lukaszo** [comment]
@k8s-bot cvm gce e2e test this #33867

**Comment by lukaszo** [comment]
@k8s-bot unit test this

Looks like downloading etcd failed

**Comment by lukaszo** [comment]
@kargakis can you one more time add lgtm? I had to rebase to get a fix for this flake #34547

**Comment by kargakis** [comment]
@lukaszo sorry, you need one more rebase

**Comment by lukaszo** [comment]
@kargakis rebased

**Comment by ahakanbaba** [comment]
HI @lukaszo  we are also very interested in benefiting from this change. Thank you for taking this on and seeing it through. 
We were wondering, is there any indication in which api version this change will be a part of. 

I see that the extensions_v1beta1.json file has changed. Does that mean that this change will be in api version v1beta1. 

This change is not in the 1.4.3 release. I also cannot see it in the v1.5.0-alpha.1 tag. 

Does anyone know which kubernetes release is likely to have this new api addition ?

**Comment by lukaszo** [comment]
Hi @ahakanbaba currently it's available only in master brach. It will be also available in v1.5.0-alpha2 release which should be released in a few days.

Yes, it will be in  api version v1beta1.

**Comment by ahakanbaba** [comment]
@lukaszo Thank you very much for the clarification.

**Pull request merged**